service/history/chasm/component.go

```diff
@@ -0,0 +1,71 @@
+package chasm
+
+import "context"
+
+type Component interface {
+	LifecycleState() LifecycleState
+
+	// TBD: the framework can just put the component in terminated state
+	// component lifecycle state can still be running when getting terminated
+	// but framework will use some rule to block incoming operations
+	// Terminate()
+
+	// we may not need this in the beginning
+	mustEmbedUnimplementedComponent()
+}
+
+// Embed UnimplementedComponent to get forward compatibility
+type UnimplementedComponent struct{}
+
+func (UnimplementedComponent) LifecycleState() LifecycleState {
+	return LifecycleStateUnspecified
+}
+
+// func (UnimplementedComponent) Terminate() {}
+
+func (UnimplementedComponent) mustEmbedUnimplementedComponent() {}
+
+// Shall it be named ComponentLifecycleState?
+type LifecycleState int
+
+const (
+	LifecycleStateUnspecified LifecycleState = 0
```

Comment:
```suggestion
	LifecycleStateUnspecified = LifecycleState(0)
```
and then it can be moved to the end of next `const` blocl.